### PR TITLE
FlexCharge: Adding authorize-capture functionality

### PR DESCRIPTION
**Pull request by Heavyblade**
## Summary:

Changes FlexCharge to add support for the authorize and capture operations and also a fix a bug on the `address_names` method

[SER-1337](https://spreedly.atlassian.net/browse/SER-1337) 
[SER-1324](https://spreedly.atlassian.net/browse/SER-1324)

### Remote Test:
Finished in 72.283834 seconds.
19 tests, 54 assertions, 1 failures, 0 errors, 0 pendings, 1 omissions, 0 notifications
94.4444% passed

### Unit Tests:
Finished in 43.11362 seconds.
5944 tests, 79908 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

### RuboCop:
798 files inspected, no offenses detected

### CODE REVIEW

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 1]
```diff
@@ -17,38 +17,61 @@ class FlexChargeGateway < Gateway
...
-      SUCCESS_MESSAGES = %w(APPROVED CHALLENGE SUBMITTED SUCCESS PROCESSING).freeze
+      SUCCESS_MESSAGES = %w(APPROVED CHALLENGE SUBMITTED SUCCESS PROCESSING CAPTUREREQUIRED).freeze
 
       def initialize(options = {})
         requires!(options, :app_key, :app_secret, :site_id, :mid)
         super
       end
 
       def purchase(money, credit_card, options = {})
+        options[:transactionType] = 'Purchase'
```
Comment (sinourain) [inline lib/active_merchant/billing/gateways/flex_charge.rb:32]: Observation: When you call the `authorize` method, `options[:transactionType]` it is `'Authorization'` but then it is called within the `authorize` a `purchase` and in this line the `options[:transactionType]` field is overwritten

```suggestion
        options[:transactionType] ||= 'Purchase'
```

File: test/remote/gateways/remote_flex_charge_test.rb [hunk 2]
```diff
@@ -139,6 +158,16 @@ def test_successful_refund
...
   end
 
+  def test_successful_void
+    @cit_options[:phone] = '998888'
+    set_credentials!
+    response = @gateway.authorize(@amount, @credit_card_mit, @cit_options)
+    assert_success response
+
+    assert void = @gateway.void(@amount, response.authorization)
+    assert_success void
```
Comment (sinourain) [inline test/remote/gateways/remote_flex_charge_test.rb:168]: When I run this test on my local environment I get this error, should I take anything else into account to perform this test?

```
  "errors"=>
         [{"item1"=>"Failed to refund order",
           "item2"=>"REFUND_FAILED",
           "item3"=>"RefundFailed",
           "item4"=>true}],
```

**Review by sinourain — APPROVED** [review]
LGTM, great job, just a couple of comments :)

File: lib/active_merchant/billing/gateways/flex_charge.rb [hunk 3]
```diff
@@ -17,38 +17,61 @@ class FlexChargeGateway < Gateway
...
+          currency: currency
+        }
+
+        commit(:capture, post, authorization)
+      end
+
       def refund(money, authorization, options = {})
-        commit(:refund, { amountToRefund: (money.to_f / 100).round(2) }, authorization)
+        order_id, _currency = authorization.split('#')
+        commit(:refund, { amountToRefund: (money.to_f / 100).round(2) }, order_id)
```
Comment (naashton) [inline lib/active_merchant/billing/gateways/flex_charge.rb:66]: Is there a reason to format the `amountToRefund` here as opposed to setting the `self.money_format` and calling the `localized_amount` method to format it?

**Review by naashton — APPROVED** [review]
LGTM, just one question about the amount formatting.

File: test/remote/gateways/remote_flex_charge_test.rb (hunk 2 above)
Comment (Heavyblade) [reply test/remote/gateways/remote_flex_charge_test.rb:168]: Ahhh yes void 'should' exactly like a refund but FlexCharge hasn't activated it yet

File: lib/active_merchant/billing/gateways/flex_charge.rb (hunk 3 above)
Comment (Heavyblade) [reply lib/active_merchant/billing/gateways/flex_charge.rb:66]: Nop, I wasn't aware and  just saw a rounding operation and did it, going to add the suggested change thanks 👍

**Pull request merged**